Add Meson Build Support

ejoerns opened the pull request:
This adds support for building RAUC with meson build system.

It also adapts CI tests to use meson instead of autotools.

Fixes: #829

jluebbe commented:
@ejoerns What needs to be done before removing the draft status?

ejoerns commented:
> @ejoerns What needs to be done before removing the draft status?

I left it on draft to possibly get some high-level feedback about possible major things I missed but maybe making it an actual PR is more appropriate for getting feedback.

jluebbe commented on meson_options.txt:
```diff
@@ -0,0 +1,67 @@
+  'create',
+  type : 'boolean',
+  value : 'true',
+  description : 'Disable bundle creation and modification commands')
```
```suggestion
  description : 'Enable/Disable bundle creation and modification commands')
```

jluebbe commented on meson_options.txt:
```diff
@@ -0,0 +1,67 @@
+  type : 'string',
+  value : 'nobody',
+  description : 'unpriviledged user for the streaming subprocess')
+
+
+
```
Fewer empty lines?

ejoerns replied:
done and replaced by separating comments

jluebbe commented on qemu-test:
```diff
@@ -45,9 +45,6 @@ check_kernel() {
 
 check_kernel || exit 1
 
-# make sure the tests are built
-make TESTS= check > /dev/null
-
```
Why can't we adapt this to meson as well?

ejoerns replied:
Tests are currently either build or not by default, depending on whether "-Dtests=true" was configured or not. Thus it is less likely to have sources built without having tests built.

One could maybe invoke meson/ninja from within the script, but then we also need to know where the build dir is which is already a bit hacky solved in `qemu-test-init`.

jluebbe replied:
Don't we need to know the build dir anyway to find the binaries?

jluebbe replied:
The intention was to avoid building in qemu (which is potentially slow when running without kvm). As we hardcode `BUILD_DIR` anyway, we could continue to build the test binaries in `qemu-init`.

jluebbe reviewed (commented):
We should probably keep a/some GitHub workflow(s) to check that autotools still work (until we actually remove it).

ejoerns commented:
> We should probably keep a/some GitHub workflow(s) to check that autotools still work (until we actually remove it).

I removed the autotools workflows in the first round since I was not sure at this point if we wanted to keep autotools or not. But if we keep it we should also keep some of the workflows, yes.

jluebbe commented:
> > We should probably keep a/some GitHub workflow(s) to check that autotools still work (until we actually remove it).
> 
> I removed the autotools workflows in the first round since I was not sure at this point if we wanted to keep autotools or not. But if we keep it we should also keep some of the workflows, yes.

I think we should keep it for one release, at least.

jluebbe commented on meson.build:
```diff
@@ -84,7 +89,10 @@ c_warn_flags = '''
-add_project_arguments(c_warn_flags, language : 'c')
+c_flags += cc.get_supported_arguments(c_warn_flags)
+c_flags += ['-fdata-sections', '-ffunction-sections', '-fno-strict-aliasing']
+c_flags += ['-Wl,--gc-sections', '-Wl,-Map,$@.map']
```
These are LD flags

jluebbe commented on qemu-test-init:
```diff
@@ -11,19 +11,12 @@ mount -t tracefs tracefs /sys/kernel/tracing/ || true
 for x in $(cat /proc/cmdline); do
   if [ "$x" = "shell" ]; then
     SHELL=1
-  elif [ "$x" = "coverage" ]; then
-    MAKE_TARGET="check-code-coverage"
-  elif [ "$x" = "distcheck" ]; then
-    MAKE_TARGET="distcheck"
```
We shouldn't remove these util we drop autotools support.

ejoerns replied:
We do not call `make` anymore in the script, thus these are useless anyway

jluebbe commented on .github/workflows/scan-build.yml:
```diff
@@ -22,14 +22,13 @@ jobs:
+    - name: Run meson
       run: |
-        ./autogen.sh
+        meson setup build -Dtests=true
```
Tests are enabled by default.

jluebbe commented on .github/workflows/tests.yml:
```diff
@@ -14,28 +14,24 @@ jobs:
-        make -j4
-        make -j4 check TESTS=
+        rm -rf build/
+        meson setup -Dservice=true -Dgpt=enabled -Dtests=true -Dwerror=true build
```
Tests are enabled by default. Check below as well.

jluebbe commented on .github/workflows/tests.yml:
```diff
@@ -127,30 +116,34 @@ jobs:
-        docker exec -e DEBIAN_FRONTEND='noninteractive' -i stable apt-get install -qy build-essential automake libtool libglib2.0-dev libcurl3-dev libssl-dev libjson-glib-dev libdbus-1-dev libfdisk-dev libnl-genl-3-dev squashfs-tools
+        docker exec -e DEBIAN_FRONTEND='noninteractive' -i stable apt-get install -qy build-essential meson libtool libglib2.0-dev libcurl3-dev libssl-dev libjson-glib-dev libdbus-1-dev libfdisk-dev libnl-genl-3-dev squashfs-tools
+
+    - name: Install meson > 0.45 for ubuntu 18.04
+      if: matrix.release == '18.04'
+      run: |
+        docker exec -e DEBIAN_FRONTEND='noninteractive' -i stable apt-get install -qy software-properties-common
+        docker exec -i stable add-apt-repository ppa:team-xbmc/ppa
+        docker exec -i stable apt-get update
+        docker exec -e DEBIAN_FRONTEND='noninteractive' -i stable apt-get install -qy meson
```
We should just disable meson tests for 18.04.

jluebbe commented on meson.build:
```diff
@@ -4,10 +4,9 @@ project(
   version : '1.8',
   meson_version : '>=0.51',
   default_options: [
-    'c_std=c99',
```
We could set 'c_std=gnu11' to pin the standard version.

jluebbe replied:
In https://github.com/rauc/rauc/compare/4478e169d2ab51e18a47b58339496585e3a16dbe..5d566674bd024218bdd06bde7ab36a44ffa2b32b#diff-30d8f6be6320feeacf686be94f48c70869b52630e01ea625f0f15adc0d57c3e4L7, this was removed again?

ejoerns replied:
Oups, rebase fail. Fixed.

jluebbe commented on meson.build:
```diff
@@ -0,0 +1,218 @@
+  output : 'config.h',
+  configuration : conf
+)
+add_project_arguments('-include' + meson.current_build_dir() / 'config.h', language : 'c')
```
Can't we use the file object here?

ejoerns replied:
I attempted to do so but this does not support obtaining the (absolute) file name as a string (which would be required here)

ejoerns commented:
I've addressed the feedback so far

jluebbe commented on README.rst:
```diff
@@ -172,16 +172,26 @@ Building from Sources
 .. note:: RAUC is intended to be built both as a host tool as well as a target
-   tool (service). Therefore it is fully prepared for `automake cross-compilation
-   <https://www.gnu.org/software/automake/manual/html_node/Cross_002dCompilation.html>`_
+   tool (service). Therefore it is fully prepared for meson.
```
```suggestion
   tool (service). Therefore it is fully prepared for cross-compilation with meson.
```

jluebbe commented on README.rst:
```diff
@@ -172,16 +172,26 @@ Building from Sources
+    meson setup build
+    ninja -C build
+
+.. note:: At the moment also `automake cross-compilation
```
```suggestion
.. note:: At the moment, also `automake cross-compilation
```

jluebbe commented on README.rst:
```diff
@@ -172,16 +172,26 @@ Building from Sources
+
+.. note:: At the moment also `automake cross-compilation
+   <https://www.gnu.org/software/automake/manual/html_node/Cross_002dCompilation.html>`_
+   is still supported, but the plan is to deprecate this in the new future
```
```suggestion
   is still supported, but the plan is to deprecate this soon.
```

ejoerns commented:
I've fixed some aspects of the scan-build invocation now.

* Added missing dependency of executable target 'rauc' to `dbus_sources` to ensure having them always generated before using them
* added missing comma-separation of `-include` and `/path/to/config.h` in `add_project_arguments()` which seems to have caused ignorance of config.h in scan-build run
* use meson's built-in `scan-build` target
* add wrapper script `scan-build-wrapper.sh` to allow disabling `unix.Malloc` checker and injecting `--status-bugs` for meson scan-build target

ejoerns commented:
Timeout of `rauc.t` seems to have been insufficient. Increased to `300` now. Next try...

jluebbe commented on scan-build-wrapper.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+scan-build -v --status-bugs -disable-checker unix.Malloc "$@"
```
Why do we want to disable `unix.Malloc`?

ejoerns replied:
because scan-build does not understand auto-cleanup. This has been disabled before, too (in the workflow file).

jluebbe reviewed (changes requested):
The "meson: add tests" commit is still worded as if tests were disabled by default. LGTM otherwise.

ejoerns commented:
> The "meson: add tests" commit is still worded as if tests were disabled by default. LGTM otherwise.

I have adapted the commit message.

The only thing missing from my perspective would be coverage support, but I could also add this in a follow-up PR.
